### ISSUE TITLE
renderToPipeableStream examples

### ISSUE THREAD

**Issue by ignatiqq** [theKashey/used-styles#48]
Hello, could you help me, where can I find an example with renderToPipeableStream?

Thank you.

**Comment by theKashey** [comment]
standard `TransformStream` approach should be able to handle it, but I've got no proofs and I've got no example as well.

**Comment by ignatiqq** [comment]
@theKashey Ok thanks. But should support for used styles also be responsive 18? I'm having a problem importing all my styles (that aren't used there) into a specific page.
Do I understand correctly that used-styles should only pick up styles from the stream or not?

**Comment by theKashey** [comment]
I've never tried. You probably should look at [interleaved stream rendering](https://github.com/theKashey/used-styles#interleaved-stream-rendering) configured _after_ `pipeableStream`, but as I've mentioned - there is no example.

if you can create such an example - that would just great and should give me a good base to start from, and may be resolve the problem for you.

**Comment by ignatiqq** [comment]
This works as expected, thanks

**Comment by ignatiqq** [comment]
I have seen the implementation and it should work as expected, but how should it work with partial hydration when we need to remove the <link> or <style> from the dom so that there are no rendering hydration issues.

**Comment by ignatiqq** [comment]
It works as expected but I have another problem related to my css loader which is separating one component style from another separated by React.Lazy and the styles used are loading it, how do we avoid this using css loader with separated fragments? How can we fix it outside the library?

**Comment by theKashey** [comment]
Can you clarify the problem - is it about ignoring some styles or some CSS files?
Take a look at https://github.com/theKashey/used-styles/blob/master/src/operations.ts#L29

**Comment by theKashey** [comment]
In that case - "thanks, but yes". This is a code from a real project reducing "used styles" to the "factually used" chunks
> where importAssets is `import { importAssets } from "webpack-imported";`
```tsx
                 return (chunks: string[]) => {
                 // get import details for given stats and given chunks
		const files = importAssets(statsConfiguration, chunks);
                 // CSS files expected to be loaded
		const stylesToLoad = new Set(files.styles.load);

		// reduce used-styles to actually "used" ones.
		return alterProjectStyles(styleData, {
			filter(name) {
                                 // will reduce original set to contain styles only from given data
				return stylesToLoad.has(name);
			}			
		});
```

**Comment by ignatiqq** [comment]
Thank you. Perhaps I don't understand something or I lack context (sorry), but is this function compatible with createStylesStream (with processing of each chunk, the styles of which we know only after another pipe from renderToPipeableStream)

**Comment by theKashey** [comment]
😅🫠 no it's not. Look like you need a way to control what is in ast in a little bit dynamic way.

So:
- "how" - exactly like in the example provided by me - just "add"/"remove" some "ast nodes". Probably add as you discover new chunks as being used.
- "how" - exactly like in the example provided by you - an option to continuously control what to add

**Comment by ignatiqq** [comment]
Thanks a lot. I would really appreciate it if you could explain when you are using this code (during stream) and what the string[] pieces are (from where). I don't understand this concept.

**Comment by theKashey** [comment]
Tell me how(when?) you track loaded chunks and I'll manage the rest.

**Comment by ignatiqq** [comment]
Thank you. As far as I understand, I can only track loaded chunks in the callback (if I use stream) in createStyleStream(second parameter) when "getUsedStylesIn" has already been executed for a specific chunk. Before this, I don't know what styles are used on the page.

**Comment by theKashey** [comment]
Not entirely what I've asked.
- you want to load styles only from "used chunks"
- this is possible if one "enable/disable" ast nodes for given chunks in a real-time. Something not possible right now, but we only need a direction.
- so how you can know which chunks are getting loaded?
  - different solutions use different tracking, but not all of them gives you access to this information during streaming.
  - according to the code you've shared - you have list of files you care about, that should be enough

Sounds like all required is to make configuration a little less static to let you "unblock" new chunks as they are getting used.

**Comment by ignatiqq** [comment]
I feel stupid now (I'm not very good at parsing and don't quite understand what "enable/disable" ast nodes means). Can you please check my mini-example repo to make sure exactly what problem we are trying to solve?

https://github.com/ignatiqq/used-style-css-modules-example

please check the "/lazy" route there and note that all styles from the project are loaded due to the extracted common style for others.

**Comment by ignatiqq** [comment]
I feel stupid now (I'm not very good at parsing and don't quite understand what "enable/disable" ast nodes means)(maybe I can read about this somewhere?). Can you please check my mini-example repo to make sure exactly what problem we are trying to solve?

https://github.com/ignatiqq/used-style-css-modules-example

please check the "/lazy" route there and note that all styles from the project are loaded by dividing the shared styles of one component into different chunks.

**Comment by theKashey** [comment]
And it short - that would not work. 
You need to "know" which lazy components got rendered, which chunks they "require" and which chunks you need to consider using for used-styles.

I'll adjust your example to work as needed, give me some time.

**Comment by ignatiqq** [comment]
@theKashey Thanks for the examples and mentioning the "webpack-imported" library. I have solved this problem on my own.

But I have one more small question. There are 3 files in my project and in each file there is a mention of h1 class. I also have a renderToPipeableStream logic. 
So, I'm facing an issue (or maybe not) related to the "getUnusableStyles" utility. 

When we sending the first html chunk to the "used-styles" transform stream (piping), on the first (every) iteration, it checks if we have unusable styles and puts all my files in the cache (as unusable). 
As far as I can see, in "src/parser/toAst" we are mapping pieces of selectors, and piece should be a class to be there.  Is this expected behavior? And is it something that can be controlled? 
I really appreciate your help 🙏.

**Comment by ignatiqq** [comment]
I really appreciate your help. I've already done this myself. Thanks for mentioning 'webpack-imported'.
